agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
+                DeviceUsageStatus::Free => {
+                    free_instances.insert(instance_name.clone());
+                }
+                DeviceUsageStatus::ReservedByConfiguration(vdev_id) => {
+                    reserved_instances.insert(instance_name.clone());
+                    vdev_ids_to_report.insert(vdev_id);
+                }
+                _ => (),
+            };
+        }
+    }
+
+    // Decide virtual device id, report already reserved virtual devices + 1 free device usage (if available) per instance
+    let mut id_to_set: u32 = 0;
+    for _x in 0..free_instances.len() {
+        while vdev_ids_to_report.get(&format!("{}", id_to_set)).is_some() {
+            id_to_set += 1;
+        }
+        let id_string = format!("{}", id_to_set);
+        vdev_ids_to_report.insert(id_string);
+        id_to_set += 1;
+    }
+
+    vdev_ids_to_report
+}
+
+async fn get_virtual_device_resources(
+    requested_vdev_ids: Vec<String>,
+    instance_config: &InstanceConfig,
+    node_name: &str,
+    instance_namespace: &str,
+    capacity: &i32,
+    kube_interface: Arc<impl KubeInterface>,
+) -> Result<HashMap<String, (String, String)>, Status> {
+    let mut device_usage_states = HashMap::new();
+    for instance_name in instance_config.instances.keys() {
+        let device_usage_state = get_instance_device_usage_state(
+            node_name,
+            instance_name,
+            instance_namespace,
+            capacity,
+            kube_interface.clone(),
+        )
+        .await;
+        device_usage_states.insert(instance_name.to_string(), device_usage_state);
+    }
+
+    let mut usage_ids_to_use = HashMap::<String, (String, String)>::new();
+
+    // Get all available virtual devices, group by device usage status,
+    // a virtual device is available if it's Free or ReservedByConfiguration with requested vdev_ids
+    let mut free_device_usage_states = HashMap::new();
+    let mut reserved_by_configuration_usage_states = HashMap::new();
+    for (instance_name, device_usage_state) in device_usage_states {
+        for (id, state) in device_usage_state {
+            match state {
+                DeviceUsageStatus::Free => {
+                    free_device_usage_states
+                        .entry(instance_name.to_string())
+                        .or_insert(HashSet::new())
+                        .insert(id);
+                }
+                DeviceUsageStatus::ReservedByConfiguration(vdev_id) => {
+                    if requested_vdev_ids.contains(&vdev_id) {
+                        reserved_by_configuration_usage_states
+                            .entry(instance_name.to_string())
+                            .or_insert(HashMap::new())
+                            .insert(id, vdev_id);
+                    }
+                }
+
+                _ => (),
+            };
+        }
+    }
+
+    // Find (Instance, usage slot) for vdev ids, (Instance, usage slot) in free_device_usage_states and
```

Comment:
i find it a bit confusing to see id, device, vdev, and slot ... i'm not certain what is equivalent (it looks like what is refered to as `id` above is `slot` below) and what is different.  it would be helpful (if there are synonyms) to use just one.

Reply:
use consistent name